src/Xamarin.Android.Build.Tasks/Utilities/XAAssemblyResolver.cs

```diff
@@ -0,0 +1,333 @@
+			if ((assemblyFile = SearchDirectory (name.Name, dir)) != null) {
+				AssemblyDefinition? loaded = Load (arch, assemblyFile);
+				if (Array.Equals (loaded?.Name.MetadataToken, name.MetadataToken)) {
+					return loaded;
+				}
+				candidate = candidate ?? loaded;
+			}
+		}
+
+		if (candidate != null)
+			return candidate;
+
+		throw CreateLoadException (name);
+	}
+
+	static FileNotFoundException CreateLoadException (AssemblyNameReference name)
+	{
+		return new FileNotFoundException ($"Could not load assembly '{name}'.");
+	}
+
+	static string? SearchDirectory (string name, string directory)
+	{
+		if (Path.IsPathRooted (name) && File.Exists (name)) {
+			return name;
+		}
+
+		var file = Path.Combine (directory, $"{name}.dll");
+		if (File.Exists (file)) {
+			return file;
+		}
+
+		return null;
+	}
+
+	public virtual AssemblyDefinition? Load (AndroidTargetArch arch, string filePath, bool forceLoad = false)
+	{
+		string name = Path.GetFileNameWithoutExtension (filePath);
+		AssemblyDefinition? assembly;
+		if (!forceLoad && cache.TryGetValue (name, out CacheEntry? entry)) {
+			assembly = SelectAssembly (arch, name, entry, loading: true);
+			if (assembly != null) {
+				return assembly;
+			}
+		}
+
+		try {
+			assembly = ReadAssembly (filePath);
+		} catch (Exception e) when (e is FileNotFoundException || e is DirectoryNotFoundException) {
+			// These are ok, we can return null
+			return null;
+		}
+
+		if (!cache.TryGetValue (name, out entry)) {
+			entry = new CacheEntry (log, filePath, assembly, arch);
+			cache.Add (name, entry);
+		} else {
+			entry.Add (arch, assembly);
+		}
+
+		return assembly;
+	}
+
+	AssemblyDefinition ReadAssembly (string filePath)
+	{
+		bool haveDebugSymbols = loadDebugSymbols && File.Exists (Path.ChangeExtension (filePath, ".pdb"));
+		var loadReaderParams = new ReaderParameters () {
+			ApplyWindowsRuntimeProjections  = readerParameters.ApplyWindowsRuntimeProjections,
+			AssemblyResolver                = this,
+			MetadataImporterProvider        = readerParameters.MetadataImporterProvider,
+			InMemory                        = readerParameters.InMemory,
+			MetadataResolver                = readerParameters.MetadataResolver,
+			ReadingMode                     = readerParameters.ReadingMode,
+			ReadSymbols                     = haveDebugSymbols,
+			ReadWrite                       = readerParameters.ReadWrite,
+			ReflectionImporterProvider      = readerParameters.ReflectionImporterProvider,
+			SymbolReaderProvider            = readerParameters.SymbolReaderProvider,
+			SymbolStream                    = readerParameters.SymbolStream,
+		};
+		try {
+			return LoadFromMemoryMappedFile (filePath, loadReaderParams);
+		} catch (Exception ex) {
+			log.LogWarning ($"Failed to read '{filePath}' with debugging symbols. Retrying to load it without it. Error details are logged below.");
+			log.LogWarning ($"{ex.ToString ()}");
+			loadReaderParams.ReadSymbols = false;
```

Comment:
though wrt my previous comment about using `parameters ?? this.readerParameters`, this presents a thread-safety concern (shared data, fun!).  (Even though you can't use shared Cecil across threads, so this shouldn't be possible, I still look at it with side-eyes.)

Reply:
I'm not sure what you mean? `loadReaderParams` is a local variable, a new instance of `ReaderParameters` created on each call to `ReadAssembly` and `ReadSymbols` is a `bool` property, I don't think this is thread-unsafe.